crates/parser/src/ast.rs

```diff
@@ -386,6 +389,327 @@ impl Spanned for ContractStmt {
     }
 }
 
+impl fmt::Display for Module {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        for stmt in self.body.iter() {
+            writeln!(f, "{}", stmt)?;
+        }
+
+        Ok(())
+    }
+}
+
+impl fmt::Display for ModuleStmt {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            ModuleStmt::Pragma(node) => write!(f, "{}", node.kind),
+            ModuleStmt::Import(node) => write!(f, "{}", node.kind),
+            ModuleStmt::TypeAlias(node) => write!(f, "{}", node.kind),
+            ModuleStmt::Contract(node) => write!(f, "{}", node.kind),
+            ModuleStmt::Struct(node) => write!(f, "{}", node.kind),
+        }
+    }
+}
+
+impl fmt::Display for Pragma {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "pragma {}", self.version_requirement.kind)
+    }
+}
+
+impl fmt::Display for Import {
+    fn fmt(&self, _f: &mut Formatter<'_>) -> fmt::Result {
+        todo!()
+    }
+}
+
+impl fmt::Display for TypeAlias {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "type {} = {}", self.name.kind, self.typ.kind)
+    }
+}
+
+impl fmt::Display for Contract {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        writeln!(f, "contract {}:", self.name.kind)?;
+        if !self.fields.is_empty() {
+            writeln!(indented(f), "{}\n", node_line_joined(&self.fields))?;
+        }
+        if !self.body.is_empty() {
+            writeln!(indented(f), "{}\n", line_joined(&self.body))?;
+        }
+
+        Ok(())
+    }
+}
+
+impl fmt::Display for Struct {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        writeln!(f, "struct {}:", self.name.kind)?;
+        writeln!(indented(f), "{}", node_comma_joined(&self.fields))
+    }
+}
+
+impl fmt::Display for TypeDesc {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            TypeDesc::Unit => write!(f, "()"),
+            TypeDesc::Base { base } => write!(f, "{}", base),
+            TypeDesc::Array { typ, dimension } => write!(f, "{}[{}]", typ.kind, dimension),
+            TypeDesc::Tuple { items } => write!(f, "({})", node_comma_joined(&items)),
+            TypeDesc::Generic { base, args } => {
+                write!(f, "{}<{}>", base.kind, comma_joined(&args.kind))
+            }
+        }
+    }
+}
+
+impl fmt::Display for GenericArg {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self {
+            GenericArg::TypeDesc(node) => write!(f, "{}", node.kind),
+            GenericArg::Int(node) => write!(f, "{}", node.kind),
+        }
+    }
+}
+
+impl fmt::Display for SimpleImportName {
```

Comment:
I just left the imports out for now, since we may not be settled on their design.